subsys/bluetooth/host/l2cap.c

```diff
@@ -67,11 +67,32 @@ NET_BUF_POOL_FIXED_DEFINE(disc_pool, 1,
 #define l2cap_lookup_ident(conn, ident) __l2cap_lookup_ident(conn, ident, false)
 #define l2cap_remove_ident(conn, ident) __l2cap_lookup_ident(conn, ident, true)
 
-struct data_sent {
-	uint16_t len;
+struct l2cap_tx_data {
+	int sent;
+	uint16_t cid;
+	bt_conn_tx_cb_t cb;
+	void *user_data;
+};
+
+struct l2cap_tx_meta {
+	struct l2cap_tx_data *data;
 };
 
-#define data_sent(buf) ((struct data_sent *)net_buf_user_data(buf))
+static struct l2cap_tx_data l2cap_tx[CONFIG_BT_CONN_TX_MAX];
+K_FIFO_DEFINE(free_l2cap_tx);
+
+static struct l2cap_tx_data *alloc_tx_meta(void)
+{
+	return k_fifo_get(&free_l2cap_tx, K_FOREVER);
```

Comment:
Can we make this `K_NO_WAIT`?

```suggestion
	return k_fifo_get(&free_l2cap_tx, K_NO_WAIT);
```

Reply:
Yes, can do to be safe not to cause any deadlock. Worst case, the application has to retry later.

I initially did these like in `conn_tx_alloc()`, but the comment there did not seem true for the deallocation in L2CAP and ATT.
```c
	/* The TX context always get freed in the system workqueue,
	 * so if we're in the same workqueue but there are no immediate
	 * contexts available, there's no chance we'll get one by waiting.
	 */
	if (k_current_get() == &k_sys_work_q.thread) {
		return k_fifo_get(&free_tx, K_NO_WAIT);
	}
```

Reply:
Ok. Will make them `K_NO_WAIT`.